### PR TITLE
Staff-Logs: Added Search for Officer and Grade Check for Leaderboard and Search

### PR DESCRIPTION
**Pull request by LeSiiN**
- done with frontend ( backend needed only)
- added LogPerms for Leaderboard and Officer Search with Notify(QBCore)

### CODE REVIEW

File: ui/style.css [hunk 1]
```diff
@@ -1729,6 +1729,20 @@ span.tag-input[contenteditable]:empty::before {
     animation: spin 2s linear infinite;
 }
 
+.leaderboard-loader {
+    margin-top: 5vh;
+    margin-left: auto;
+    margin-right: auto;
```
Comment (xFutte) [inline ui/style.css]: This can actually be refactored to:
```suggestion
    margin: 5vh auto 0 auto;
```

File: ui/style.css [hunk 2]
```diff
@@ -5095,6 +5109,126 @@ span.tag-reports-input[contenteditable]:empty::before {
     user-select: none;
 }
 
+.leaderboard-officer-title {
+    background-color: var(--color-4);
+    color: white;
+    font-size: 18px;
+    font-weight: bolder;
+    padding-top: 5px;
+    padding-bottom: 5px;
```
Comment (xFutte) [inline ui/style.css]: Can be refactored to `padding: 5px 0;`

File: ui/style.css [hunk 3]
```diff
@@ -5095,6 +5109,126 @@ span.tag-reports-input[contenteditable]:empty::before {
...
+    background-color: var(--color-4);
+    color: white;
+    font-size: 18px;
+    font-weight: bolder;
+    padding-top: 5px;
+    padding-bottom: 5px;
+    text-align: center;
+    margin: auto;
+    margin-top: 10px;
+    margin-bottom: 0px;
```
Comment (xFutte) [inline ui/style.css]: Refactor to `margin: 10px auto 0 auto`

File: ui/style.css [hunk 4]
```diff
@@ -5095,6 +5109,126 @@ span.tag-reports-input[contenteditable]:empty::before {
...
+
+.leaderboard-title:hover {
+    background-color: var(--color-3);
+}
+
+.leaderboard-search-input {
+    background-color: var(--color-3);
+    color: white;
+    padding-top: 10px;
+    padding-bottom: 10px;
```
Comment (xFutte) [inline ui/style.css:5131]: Refactor to `padding: 10px auto`

File: ui/style.css [hunk 5]
```diff
@@ -5095,6 +5109,126 @@ span.tag-reports-input[contenteditable]:empty::before {
...
+    text-align: center;
+    margin: auto;
+    margin-top: 10px;
+    margin-bottom: 0px;
```
Comment (xFutte) [inline ui/style.css:5135]: `margin: 10px auto 0 auto;`

File: ui/style.css [hunk 6]
```diff
@@ -5095,6 +5109,126 @@ span.tag-reports-input[contenteditable]:empty::before {
...
+    display: none;
+}
+
+#leaderboard-search-input {
+    outline: none;
+    border: none;
+    width: 95%;
+    color: white;
+    font-size: 15px;
+    background-color: rgba(0, 0, 0, 0);
```
Comment (xFutte) [inline ui/style.css]: Just use `background-color: transparent`

File: ui/style.css [hunk 7]
```diff
@@ -5095,6 +5109,126 @@ span.tag-reports-input[contenteditable]:empty::before {
...
+#leaderboard-search-input::placeholder {
+    color: white;
+}
+
+.leaderboard-items {
+    display: flex;
+    flex-direction: column;
+    overflow: auto;
+    margin: auto;
+    margin-top: 15px;
```
Comment (xFutte) [inline ui/style.css]: Combine the three margin properties to `margin: 15px auto 0 auto;`

File: ui/style.css [hunk 8]
```diff
@@ -5095,6 +5109,126 @@ span.tag-reports-input[contenteditable]:empty::before {
...
+    user-select: none;
+
+}
+
+.leaderboard-image {
+    width: 10vh;
+    height: 10vh;
+    margin-left: 10px;
+    margin-top: auto;
+    margin-bottom: auto;
```
Comment (xFutte) [inline ui/style.css:5182]: I'm not sure if this is needed.

File: ui/style.css [hunk 9]
```diff
@@ -5095,6 +5109,126 @@ span.tag-reports-input[contenteditable]:empty::before {
...
+    display: flex;
+    width: 100%;
+    height: 5px;
+    margin-top: 10px;
+    margin-bottom: 5px;
+}
+
+.leaderboard-tag {
+    margin: auto;
+    margin-left: 0px;
```
Comment (xFutte) [inline ui/style.css]: Is this needed? If so, combine the three margin properties to `margin: auto 6px auto 0px`

File: ui/style.css (hunk 1 above)
Comment (LeSiiN) [reply ui/style.css]: Did that thank you <3

File: ui/style.css (hunk 2 above)
Comment (LeSiiN) [reply ui/style.css]: Did that thank you <3

File: ui/style.css (hunk 3 above)
Comment (LeSiiN) [reply ui/style.css]: Did that thank you <3

File: ui/style.css (hunk 4 above)
Comment (LeSiiN) [reply ui/style.css:5131]: Did that thank you <3

File: ui/style.css (hunk 5 above)
Comment (LeSiiN) [reply ui/style.css:5135]: Did that thank you <3

File: ui/style.css (hunk 6 above)
Comment (LeSiiN) [reply ui/style.css]: Did that thank you <3

File: ui/style.css (hunk 7 above)
Comment (LeSiiN) [reply ui/style.css]: Did that thank you <3

File: ui/style.css (hunk 8 above)
Comment (LeSiiN) [reply ui/style.css:5182]: Did that thank you <3

File: ui/style.css (hunk 9 above)
Comment (LeSiiN) [reply ui/style.css]: Did that thank you <3

File: ui/app.js [hunk 10]
```diff
@@ -1052,6 +1053,42 @@ $(document).ready(() => {
...
+  $(".leaderboard-title").click(function () {
+    if (canSearchForOfficer == true) {
+      if ($(".leaderboard-search-input").css("display") == "none") {
+        $(".leaderboard-search-input").slideDown(250);
+        $(".leaderboard-search-input").css("display", "block");
+      } else {
+        $(".leaderboard-search-input").slideUp(250);
+        setTimeout(() => {
+          $(".leaderboard-search-input").css("display", "none");
+        }, 250);
```
Comment (xFutte) [inline ui/app.js]: If I recall correct, the `slideDown` actually takes a function as a second parameter which triggers, when the animation is done. So you might be able to do something like:
```suggestion
        $(".leaderboard-search-input").slideUp(250, () => {
            $(".leaderboard-search-input").css("display", "none");
        });
```

Try giving it a go. The code would be way more clean that way.

File: ui/app.js [hunk 11]
```diff
@@ -5736,3 +5774,75 @@ $(".map-clear").on('click', function() {
...
+  canSearchForOfficer = true;
+  $(".leaderboard-items").empty();
+
+  if (result.length < 1) {
+    $(".leaderboard-items").html(
+      `
+                      <div class="leaderboard-item" data-id="0">
+
+                          <div style="display: flex; flex-direction: column; margin-top: 2.5px; margin-left: 5px; width: 100%; padding: 5px;">
+                          <div style="display: flex; flex-direction: column;">
```
Comment (xFutte) [inline ui/app.js:5787]: This should be moved into CSS classes.

File: ui/app.js [hunk 12]
```diff
@@ -5736,3 +5774,75 @@ $(".map-clear").on('click', function() {
...
+    let charinfo = value.charinfo;
+    let metadata = value.metadata;
+
+    if (typeof value.charinfo == "string") {
+      charinfo = JSON.parse(charinfo);
+    }
+  
+    if (typeof value.metadata == "string") {
+      metadata = JSON.parse(metadata);
+    }
```
Comment (xFutte) [inline ui/app.js:5812]: Can `.metadata` return as an integer? Or what do you want to check here?

File: ui/app.js [hunk 13]
```diff
@@ -5736,3 +5774,75 @@ $(".map-clear").on('click', function() {
...
+
+    if (!metadata) {
+      metadata = {};
+    }
+
+    if (!metadata.licences) {
+      metadata.licences = {};
+    }
+  
+    let name = charinfo.firstname + " " + charinfo.lastname;
```
Comment (xFutte) [inline ui/app.js]: A better way of doing this to make it more readable would be:

```suggestion
    let name = `${charinfo.firstname} ${charinfo.lastname}`;
```

File: ui/app.js [hunk 14]
```diff
@@ -5736,3 +5774,75 @@ $(".map-clear").on('click', function() {
...
+  
+    if (value.pp == '') {
+      value.pp = 'img/not-found.webp'
+    }
```
Comment (xFutte) [inline ui/app.js:5826]: "pp" is not really that saying of what the is. If possible, I'd suggest to change it to imageUrl.

File: ui/app.js [hunk 15]
```diff
@@ -5736,3 +5774,75 @@ $(".map-clear").on('click', function() {
...
+
+    leaderboardOfficerHTML += `
+                  <div class="leaderboard-item" data-id="${value.citizenid}">
+                      <img src="${value.pp}" class="leaderboard-image">
+                        <div style="display: flex; flex-direction: column; margin-top: 2.5px; margin-left: 5px; width: 100%; padding: 5px;">
```
Comment (xFutte) [inline ui/app.js:5831]: The styling here is also better of going into a CSS class.

File: server/main.lua [hunk 16]
```diff
@@ -245,6 +245,37 @@ QBCore.Functions.CreateCallback('mdt:server:SearchProfile', function(source, cb,
 	return cb({})
 end)
 
+QBCore.Functions.CreateCallback('mdt:server:SearchOfficerProfile', function(source, cb, sentData)
+	if not sentData then  return cb({}) end
```
Comment (xFutte) [inline server/main.lua]: For debugging purposes, something like:

```suggestion
	if not sentData then  
	    print('No data received for searching after officer')
	    return cb({}) 
	end
```

Up to you if you think it makes sense.

File: ui/app.js (hunk 10 above)
Comment (xFutte) [inline ui/app.js:1065]: Basically the same as below comment.

File: server/main.lua (hunk 16 above)
Comment (LeSiiN) [reply server/main.lua]: done <3

File: ui/app.js (hunk 14 above)
Comment (LeSiiN) [reply ui/app.js:5826]: pp is profilepic, so i guess thats okay

File: ui/app.js (hunk 13 above)
Comment (LeSiiN) [reply ui/app.js]: done <3

File: ui/app.js (hunk 12 above)
Comment (LeSiiN) [reply ui/app.js:5812]: code is from profile search so someone had an idea here

File: ui/app.js (hunk 10 above)
Comment (LeSiiN) [reply ui/app.js]: done <3

**Comment by MonkeyWhisper** [comment]
This PR it's just too much right now for my eyes

Please squash and resend it :)

**Pull request closed without merging**